### PR TITLE
Add `InvariantOf` type

### LINKED ISSUE #288: Proposal: `InvariantOf`
Inspired by [InvariantOf](https://flow.org/en/docs/lang/variance/#toc-invariance).

It could be useful to make object invariant. It is similar `opaque` type without brand or tag.

```typescript
import {Opaque} from './opaque';

export type InvariantOf<Type> = Opaque<Type, keyof Type>;
```

I published this utility type as [package](https://github.com/younho9/invariant-of), but I wish to add it to this project.

[Here](https://github.com/younho9/invariant-of/blob/main/index.test-d.ts) are some test cases.

I'm willing to do PR. 😃

### PR DESCRIPTION
**Pull request by younho9**
resolve #288

### CODE REVIEW

**Comment by papb** [comment]
In (your package)[https://github.com/younho9/invariant-of], I see that you also monkey-patched the global `Object.keys` typings to behave better with your custom type. Also, I noticed that [your example Playground](https://www.typescriptlang.org/play?#code/PTAEEkDtQZwFwE4FcDGckIIYBtRwJ4AOAlpAOYA0okmCCA9gO6lmgICm6C0Bh7o9AGagA8gCMAVuzQA6ANbt8oYjFAATTOXYMkMGQCh9gpJDTF60BfhGCAggCEAFPUnS4AfQBcoBwEpQAN76oGycGNDiUrJWMM6uaO7+mKqOVkI+9r4A2gC6+gC+hqRw2oKYKPwOgcGgmN6QSAC2YtoA3DVi3vAILAVFkCUIZRUZAMKg7AAeJZBqqlVBISjeYvT02Oyaffpq0ti0-CgW8LWdGTt7B6BHkCeYYstjhlY2Do73-iCgqYrpfrn6F52Jz3FCfMA-fB-TK5GR4IjsGAoHqEODqeiIyAAcjRjFokAM+i+AFUYCwIJAAG60YiaOA2Qy7FD7DjXY5ouCYMjeEzEACOSH4MHwzXW7X0vH4UGpPTpAAUGHwEAQADwAFQRAD5QABeb60bmgDV8fw67XG9jiyUUmW0gYAZWIZBoXHY6q1uuqIQ4mDUFmwSiynLIOW80pp8sV2lVFs17Xy4qmhHoyvhfBtEYGNndfG1notoAAZBnZQ6nS6MG7Y4nJsnU8ZTHBzNBSLa6dnY45qdhBd4Lb4w1TM-TBDn2NrFqEuNBu4Laqpw6WR2O49sG2YLKAgYu7XA3i4oh5B22s6OHJr-JOONPRPE4PJFLED25EvPvmlhP88oV9DcTq3hyqPUAKXGx3jEXx2j-NEQN3BxxmAodQMEcCwXFbckLgpxYLpPxWlAL5IWhbI8gwk892wzDcPsUZIIIsBJSRFFcXxAwgA) defines an `invariantOf` helper function which I like and even think necessary for the ergonomics of your proposed type.

To me it feels like both things (`invariantOf` and monkey-patching the Object.keys typings) are necessary for this type to really shine, so maybe it should remain a separate package?

**Comment by younho9** [comment]
> Also, I noticed that [your example Playground](https://www.typescriptlang.org/play?#code/PTAEEkDtQZwFwE4FcDGckIIYBtRwJ4AOAlpAOYA0okmCCA9gO6lmgICm6C0Bh7o9AGagA8gCMAVuzQA6ANbt8oYjFAATTOXYMkMGQCh9gpJDTF60BfhGCAggCEAFPUnS4AfQBcoBwEpQAN76oGycGNDiUrJWMM6uaO7+mKqOVkI+9r4A2gC6+gC+hqRw2oKYKPwOgcGgmN6QSAC2YtoA3DVi3vAILAVFkCUIZRUZAMKg7AAeJZBqqlVBISjeYvT02Oyaffpq0ti0-CgW8LWdGTt7B6BHkCeYYstjhlY2Do73-iCgqYrpfrn6F52Jz3FCfMA-fB-TK5GR4IjsGAoHqEODqeiIyAAcjRjFokAM+i+AFUYCwIJAAG60YiaOA2Qy7FD7DjXY5ouCYMjeEzEACOSH4MHwzXW7X0vH4UGpPTpAAUGHwEAQADwAFQRAD5QABeb60bmgDV8fw67XG9jiyUUmW0gYAZWIZBoXHY6q1uuqIQ4mDUFmwSiynLIOW80pp8sV2lVFs17Xy4qmhHoyvhfBtEYGNndfG1notoAAZBnZQ6nS6MG7Y4nJsnU8ZTHBzNBSLa6dnY45qdhBd4Lb4w1TM-TBDn2NrFqEuNBu4Laqpw6WR2O49sG2YLKAgYu7XA3i4oh5B22s6OHJr-JOONPRPE4PJFLED25EvPvmlhP88oV9DcTq3hyqPUAKXGx3jEXx2j-NEQN3BxxmAodQMEcCwXFbckLgpxYLpPxWlAL5IWhbI8gwk892wzDcPsUZIIIsBJSRFFcXxAwgA) defines an `invariantOf` helper function which I like and even think necessary for the ergonomics of your proposed type.

Having actual functions here is a non-goal. This package is for types only, meaning nothing ends up compiled into actual JS code. https://github.com/sindresorhus/type-fest/issues/292#issuecomment-944272247

If `invariantOf` type is accepted, it would be possible to propose a helper function to the [ts-extras](https://github.com/sindresorhus/ts-extras).

> In (your package)[https://github.com/younho9/invariant-of], I see that you also monkey-patched the global `Object.keys` typings to behave better with your custom type.

I think it's convenient for me to monkey-patch global typing, but I think discussion is necessary.

**Comment by sindresorhus** [comment]
I think this type can be useful for your own types.

**Comment by sindresorhus** [comment]
> If invariantOf type is accepted, it would be possible to propose a helper function to the ts-extras.

I don't think we should override globals in a package, but it could maybe used to improve https://github.com/sindresorhus/ts-extras/blob/main/source/object-keys.ts

**Comment by sindresorhus** [comment]
Could this one be used to make the `Opaque` type?

File: source/invariant-of.d.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	// Make invariant function signature.
+	readonly [tag]: (argument: Type) => Type;
+};
+
+/**
+Create an [invariant type](https://basarat.gitbook.io/typescript/type-system/type-compatibility#footnote-invariance), which is a type that does not accept supertypes and subtypes.
+
+Use-case:
+- Prevent runtime errors that may occur due to assigning subtypes to supertypes.
+- Improve type signature of object methods like [`Object.keys()` or `Object.entries()`](https://github.com/microsoft/TypeScript/pull/12253#issuecomment-263132208) by sealing the object type.
```
Comment (sindresorhus) [inline source/invariant-of.d.ts:13]: Can you think of any more specific use-cases?

File: source/invariant-of.d.ts [hunk 2]
```diff
@@ -74,4 +69,4 @@ keyOfInvariantFooBar(invariantOf(fooBarBaz)); // Error: Argument of type 'Invari
 
 @category Type
 */
-export type InvariantOf<Type> = Type & InvariantSignature<Type>;
+export type InvariantOf<Type> = Opaque<Type, (argument: Type) => Type>;
```
Comment (younho9) [inline source/invariant-of.d.ts:72]: > Could this one be used to make the `Opaque` type?

Improve to use the `Opaque` type.

File: source/invariant-of.d.ts [hunk 3]
```diff
@@ -51,7 +51,7 @@ function keyOfFooBar(fooBar: FooBar) {
 }
 
 keyOfFooBar(fooBar) //=> (keyof FooBar)[]
-keyOfFooBar(fooBarBaz) // Allowed but, error at runtime
+keyOfFooBar(fooBarBaz) //=> (keyof FooBar)[] but, (keyof FooBarBaz)[] at runtime
```
Comment (younho9) [inline source/invariant-of.d.ts:54]: There was a mistake in the document here.

File: source/invariant-of.d.ts (hunk 1 above)
Comment (younho9) [reply source/invariant-of.d.ts:8]: It's hard to think of any more specific use-cases for me.

**Pull request merged**